0004/README.md

```diff
@@ -0,0 +1,514 @@
+
+Non-objectives:
+
+-  Policies for what direct or indirect package dependencies (or dependency
+    versions) to expect for an npm software package; policies for vulnerability
+    scans.
+
+## Overview
+
+Here, we highlight the main architecture elements, their roles, and limitations.
+We present more detail in subsequent sections, with additional detail in the
+Appendix.
+
+### Extended software supply chain model
+
+![Extended model](extended-model.svg)
+
+The figure shows a generic supply chain for software packages, from source repository
+to installation, extended with two SLSA provenance policy evaluation points:
+
+-  When a software package is published to a package registry.
+-  When a software package is installed from a package registry.
+
+A SLSA provenance policy contains expectations for provenance, as discussed in later
+sections.
+
+SLSA Provenance expectations may be evaluated at other points inside or outside
+the above pipeline. Examples are a pre-build provenance check for build inputs,
+or an organization-wide proxy that checks SLSA provenance for package downloads.
+SLSA Provenance expectations may also differ between development versus
+production environments, or private versus public registries. We will not cover
+these differences in detail, but expect that the use cases can use similar SLSA
+provenance expectations as described here.
+
+### SLSA provenance expectations, benefits and limitations
+
+Each SLSA provenance policy covers multiple versions for a specific package
+name. For example, the policy for package ‘foo' could expect that each covered
+version:
+
+-  Has SLSA provenance that is signed with a key that is bound to a trusted
+    identity. The key could be long-lived, rotated regularly, or ephemeral, and
+    the identity could be a builder, individual, or organization.
+-  The provenance contains claims that:
+    -  Combined together satisfy the requirements for SLSA level X.
+        -  A policy evaluator may limit the maximum SLSA level
+            depending on the builder identity or other context. 
+
+    -  The package was built from the repository
+        [https://example.com/path/to/foo](https://example.com/path/to/foo).
+
+-  Other provenance claims may be relevant for specific use cases.
+
+When enforced, these expectations will ensure that a compromise of a package
+registry upload credential (or a compromise of the registry itself) cannot
+result in the installation of a rogue ‘foo' package version that was built from
+an unexpected repository or with an unexpected builder. 
+
+If a ‘good' package version does not meet policy expectations, the expectations
+need to be updated for that package version and onwards, so that earlier package
+versions will continue to pass expectations (differnt dependencies may require
+different versions of a package).
+
+These expectations cannot detect 'upstream' compromises, for example the
```

Comment:
Sorry, why not exactly? It's not entirely clear to me right now, especially if you tie the right types of attestations together, combined with a sufficiently expressive policy.

Reply:
With the policies mentioned here, are mostly around verifying that the correct upstream repository and builder was used. There are some loose mentions of other claims too. But in the general case I think it would be hard to detect the examples listed below with just a SLSA build provenance? 
* If the builder/build system is compromised that would require deep introspection of the system, e.g. trace syscalls to see that a file is being replaced during build/packaging time etc.
* Malicious change to a source repository. This would be impossible(?) to detect during build, and would require source level attestations?

Reply:
Would it help if I add text along the lines of: "A provenance policy specifies expectations for provenance attestation properties. It can protect only the part of the supply chain after the attestation is signed. Not the parts that come before: the build process, or the source repository content."

Reply:
I updated the text so that it  now covers both a) the part of the software supply chain that the  expectations can protect, and b) the part that they cannot. protect

Reply:
@trishankatdatadog Are you OK with the text as updated?

Reply:
Sorry, been super busy, let me try to find some downtime...